### ISSUE TITLE
How do I get PGModeler to recognise new data types provided from an Extension?

### ISSUE THREAD

**Issue by PowerUser1234** [pgmodeler/pgmodeler#1628]
I am getting a PostgreSQL **extension** built that will create two new base 10 number types, and some associated functions for them.
This is a PostgreSQL **extension**, for Linux and Windows. In order to build and edit a database model in PGModeler using the two
new types, in PGModeler version 0.9.4, (built by following https://pgmodeler.io/support/installation), how would I apply this **extension** to PGModeler so that it recognises it, for database creation and editing purposes, in PGModeler?

**Comment by rkhaotix** [comment]
Let's see if I got it: are you trying to make an extension that was built for PostgreSQL to be recognized or loaded in pgModeler?

**Comment by PowerUser1234** [comment]
Yes, that it correct.  Is there just a default option or menu item in 0.9.4 PgModeler that I can use to accomplish this,
either by the extension files, or a way to forge an active connection to PostgreSQL from PgModeler?

How may I make PgModeler aware of a particular plugin, so that I can include its additions in PostgreSQL database modelling,
editing, or PostgreSQL Database syncing?

**Comment by rkhaotix** [comment]
Well, loading an PostgreSQL extension in pgModeler isn't possible. They are completely different things. A pgModeler plugin is a set of features that extends the ones already present in the tool, and they a written in C++ on top of Qt's plug-in infrastructure.
https://www.pgmodeler.io/support/docs/81-extending-features-with-plug-ins?v=0.9.4

Now a PostgreSQL extension is written in C and extends the database system features.

If you want to include the data types in pgModeler you can achieve that without using a plugin unless you want to make something more sophisticated that isn't made by the tool. Anyway, did you try to create types in pgModeler?

https://www.pgmodeler.io/support/docs/333-user-defined-types?v=0.9.4
https://www.pgmodeler.io/support/docs/714-creating-dummy-objects?v=0.9.4

**Comment by PowerUser1234** [comment]
-What I am getting created is a PostgreSQL **extension**.  No, using the **create type** statement is not enough
for the two new data types in my new extension.  It has to be developed in C.  My extension is going to have two different versions, a **Windows** one and a **Linux** one.  What I want to verify, or learn really, is how I can get PGModeler to 
recognise original PostgreSQL types that are created in an **extension** to the DBMS. So that PGModeler can synchronise,
create and edit DB models involving the extension only types, and not just balking and producing some error when it 
encounters instances of those types.  What needs to happen with PGModeler for it to understand original 
**PostgreSQL extension types**?

**Comment by rkhaotix** [comment]
You have two options:

1) If the two data types are created by different extensions you can crate both extensions in pgModeler and check the box `Handles data type`.

https://www.pgmodeler.io/support/docs/317-extensions?v=0.9.4

2) If they are created by the same extension, you can create the extension in your database model and then create two dummy data types so they can be available to be associated with columns, functions, and so on. A dummy object in pgModeler is any object that has the SQL code disabled which causes no code to be executed in the server, being useful only for modeling purposes.

An example of the situations above can be the `hstore` extension, which creates the data types `hstore` and `gstore`. But most of the time one wants to use only `hstore` in the database model. So, it is enough to create the extension `hstore` in the model and check the `Handles data type` option. 

Now, if the user wants to use `hstore` and `gstore` data types, he/she needs to create the extension `hstore` without check the option `Handles data type` and then create two types named `hstore` and `gstore` and disable their SQL code generation. The data types must be create under the same schema as the extension or the generated code may fail during execution.

![image](https://user-images.githubusercontent.com/2205476/148691315-54d37ccb-d0cb-4e7c-a60c-0f5bc31e791d.png)

[extension_example.zip](https://github.com/pgmodeler/pgmodeler/files/7835455/extension_example.zip)

**Comment by PowerUser1234** [comment]
The situation with me is that there will be **two data types in one extension**.
I have read what you have done in your example here, with your gstore and hstore types.

-My question is, how have you made pgmodeler aware of these original types?

_How do I load an extension into PgModeler so that it may read, load and aquire the new data types, for modelling
usage?_

**Comment by rkhaotix** [comment]
> -My question is, how have you made pgmodeler aware of these original types?

This is what I'm trying to explain in previous comments. These are data types created manually using the user-defined dialog:
https://www.pgmodeler.io/support/docs/333-user-defined-types?v=0.9.4

In this case, I have created them as enum types and disabled their SQL code since they don't need to have SQL code generated in the export.

> How do I load an extension into PgModeler so that it may read, load and aquire the new data types, for modelling
usage? What options and menus do I need to access, and how, please?

Again, pgModeler don't know how to "read, load and aquire" new data types installed by extensions. **You** have to create them manually on your database model. I explained everything in my previous comment.

About reverse engineering and diff, there are options that make pgModeler retrieve these data types from a database... one of them is `Import extension objects` but you may need to use  `Import system objects` in some situations.

https://www.pgmodeler.io/support/docs/52-import-a-database?v=0.9.4
https://www.pgmodeler.io/support/docs/53-compare-model-and-database?v=0.9.4

**Comment by benjamin-kirkbride** [comment]
This is being worked on here: https://github.com/pgmodeler/pgmodeler/issues/1630
